botorch/utils/stat_dist.py

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from torch import Tensor
+
+
+def mvn_kl_divergence(
+    p_mean: Tensor, q_mean: Tensor, p_covar: Tensor, q_covar: Tensor
+) -> Tensor:
+    """Computes the kl divergence between two multivariate normal distributions.
+
+    Args:
+        p_mean: batch_shape x dist_shape x 1-dim Tensor of means of first dist
+        q_mean: batch_shape x dist_shape x dist_shape-dim Tensor of covariances
+        of first dist where the covariances are (optionally) in the q-batch dim.
+        p_mean: batch_shape x dist_shape x 1-dim Tensor of means of second dist
+        q_mean: batch_shape x dist_shape x dist_shape-dim Tensor of covariances
+        of first dist where the covariances are (optionally) in the q-batch dim.
+
+    Returns:
+        Tensor: batch_shape x dist_shape kl divergence between the multivariate
+        gaussian distributions p and q
```

Comment:
Could you also explain what the "(optionally) in the q-batch dim" means. Do we have `dist_shape = q` here?

I also wonder if PyTorch has a more performant implementation of this, though I couldn't find one for MVNs on a quick look.
```suggestion
    """Computes the KL-divergence between two multivariate normal distributions.

    Args:
        p_mean: A `batch_shape x dist_shape x 1`-dim Tensor of means of
            the first distribution.
        q_mean: A `batch_shape x dist_shape x dist_shape`-dim Tensor of
            covariances of the first distribution,  where the covariances
            are (optionally) in the q-batch dim.
        p_mean: A `batch_shape x dist_shape x 1`-dim Tensor of means of
            the second distribution.
        q_mean: A `batch_shape x dist_shape x dist_shape`-dim Tensor of
            covariances of the second distribution where the covariances
            are (optionally) in the q-batch dim.

    Returns:
        A tensor of shape `batch_shape x dist_shape` denoting the KL-divergence 
        between the multivariate Gaussian distributions p and q.
```

Reply:
That mean that the q-batch can optionally be passed in in the second dimension. Basically, the **co**variance (and not the variance) within the q-batch is computed if the second dimension has shape > 1.

Reply:
I did have a good look at the default pytorch stuff, which has KL divergence in there. It's a bit of a shame to do it this way, since it's not very clean. However, I didn't find a better solution (but I'd be willing to fix a better one if you have ideas!)